### ISSUE TITLE
Twemoji Folder Structure

### ISSUE THREAD

**Issue by vijayaraghavanramanan** [cdnjs/cdnjs#8860]
Twemoji's github page has a complicated folder structure because of two versions. v1 and v2.

Right now, the javascript file is referenced as:

https://cdnjs.cloudflare.com/ajax/libs/twemoji/2.1.3/2/twemoji.min.js

Note the /2/ after /2.1.3

This is not needed and you could simply have a structure:

├──72x72
├──svg
twemoji.min.js

Thoughts?

**Comment by vijayaraghavanramanan** [comment]
In fact the package.json already points to folder 2/

https://github.com/twitter/twemoji/blob/gh-pages/package.json

**Comment by vijayaraghavanramanan** [comment]
Any update on this?

**Comment by PeterDaveHello** [comment]
Hi @WebReflection,
Is it possible to add a `dist` folder for it so that we can update it easily? Thanks.

**Comment by WebReflection** [comment]
unfortunately we have, since the beginning, a very simple 1:1 CDN automation so whatever lands in this repo, will be available through MaxCDN (so that cdnJS is not the only CDN we use).

Having V2 breaking changes, we decided to migrate the previous structure, everything at the root of the project, to a major folder structure so that on MaxCDN you can http://twemoji.maxcdn.com/1/twemoji.min.js or http://twemoji.maxcdn.com/2/twemoji.min.js and obtain the latest major version.

TL;DR this won't change or we'll break every site that uses current MaxCDN links.

Introducing another folder doesn't seem to solve the problem, it creates another folder structure inconsistency with MaxCDN and the repo but if there's a way to tell cdnJS to just use the tag to point at the specific folder, it'd be cool but pleas consider that by default assets are also looked in the respective Vx folder.

As summary, is there any real-world issue on having tag version and `2/` after, beside being not so good looking? If somebody points at the fact there are 2 extra chars I'd like to remind everyone we use `.min` on the web to specify minified versions ;-)

If there are real issues please tell me more so I can discuss eventually with MaxCDN people and also with other people involved in the project.

Thank you.

**Comment by PeterDaveHello** [comment]
@WebReflection Thanks for your super detail explanation! I got your concerns and I will think about how can we do it the better way, thanks a lot.

**Comment by vijayaraghavanramanan** [comment]
I think one can make a case against the 2/ 

This is because even cdnjs incorrectly points to the version 1 when I search on cdnjs

![twemoji](https://cloud.githubusercontent.com/assets/13736892/18250475/4cde3b88-73a1-11e6-845b-f6afd2a9372a.png)

Even if cdnjs corrects it, a user might incorrectly point it to the version 1. So it's best to not have the 2/ in the file path.

**Comment by WebReflection** [comment]
> This is because even cdnjs incorrectly points to the version 1

... can you please tell me what is that *even* about? I've noticed we haven't specified in the README how to grab the latest via cdnJS, but we clearly specified maxCDN link and also mentioned new folders in the breaking change.

https://github.com/twitter/twemoji#breaking-changes-in-v2

**Comment by vijayaraghavanramanan** [comment]
Hi @WebReflection 

> .. can you please tell me what is that even about?

If you notice the screenshot, the version incorrectly points to V1 twemoji.min.js 

Search result gives https://cdnjs.cloudflare.com/ajax/libs/twemoji/2.1.5/twemoji.min.js

as I show in my screenshot. 

The correct link as per current configuration is https://cdnjs.cloudflare.com/ajax/libs/twemoji/2.1.5/2/twemoji.min.js

Hence's the need to be clear and simple.

**Comment by WebReflection** [comment]
this is unfortunately a limit of our, surely too simple, initial setup.

We cannot remove the old V1 from the root of the repo, we cannot replace it with V2 because it breaks.

Everyone should ignore files at the root of the repo and use either `/1/twemoji.min.js` or `/2/twemoji.min.js` and either `/1/svg` or  `/2/svg`

Everything else is legacy, and as such, it won't probably change.

If there is a way to specify right folder for the main in here, I'd be happy to use such way.
I'll also double check that default assets will keep working for V2

**Comment by vijayaraghavanramanan** [comment]
Yes you've made a good case for not changing the original twemoji repo. 

Personally I am not arguing for the twemoji repo to make any change. I am only making the case for a change in cdnjs. 

I think cdnjs can still have a simple setup because your package.json correctly points to folder 2/ as main.

**Comment by WebReflection** [comment]
if I can specify a whitelist that should be included "as if it was the root" I'd love to do that and keep things as clean as possible. Please ping me if there's any solution I can try.

**Comment by PeterDaveHello** [comment]
@vijayaraghavanramanan If it's only about the point to the mainfile, I can just modify that.

**Comment by vijayaraghavanramanan** [comment]
@PeterDaveHello not the only point (if you're asking about the search result on cdnjs). The main point was about using 2/ as the base folder so that we have simple links such as https://cdnjs.cloudflare.com/ajax/libs/twemoji/2.1.5/twemoji.min.js as the main version instead of adding a 2/.

**Comment by WebReflection** [comment]
> instead of adding a 2/

I don't think anyone in this world would actually manually write a CDN link so I'm not sure, once fixed the link to point at the right file, what exactly would be the gain, specially if it requires way more effort?

**Comment by vijayaraghavanramanan** [comment]
The issue is unknowingly assuming that the link https://cdnjs.cloudflare.com/ajax/libs/twemoji/2.1.5/twemoji.min.js as the main file. 

Even cdnjs search assumed this. 

Plus I don't think there's effort from your side. Just from cdnjs's side pointing to the folder 2 of twemoji's git as the main folder.

**Comment by PeterDaveHello** [comment]
@vijayaraghavanramanan I just updated it to be https://cdnjs.cloudflare.com/ajax/libs/twemoji/2.1.5/2/twemoji.min.js

**Comment by PeterDaveHello** [comment]
I don't want to have any breaking changes here either, so maybe we just leave the other parts alone until v3? Though the users will be affected may actually use a wrong version / file.

**Comment by PeterDaveHello** [comment]
I just found another potential issue here, there is a base path inside the javascript:

```js
var twemoji={base:"https://twemoji.maxcdn.com/2/
```
I haven't trace the code yet, not sure if it hurts ...

**Comment by vijayaraghavanramanan** [comment]
@PeterDaveHello 

Not an issue really. The way it works is that if no settings are passed to twemoji, it will assume the base and other settings such as folder/file type for the emoji (such as png or svg). The user can pass cdnjs as the base and the code will download emoji png/svg files from cdnjs instead, if the webpage needs it.

**Comment by PeterDaveHello** [comment]
@vijayaraghavanramanan The issue for me here is I'm afraid of the users may not know that they need to manually configure it so CDNJS is just the entry and then use another CDN to load the files on the background, this can not benefit from http/2.0 on the same domain and make users confused about the connections/performance.

**Comment by WebReflection** [comment]
@PeterDaveHello how would you understand at runtime the right CDN to use from a script loaded either via HTML or JS? from JS point of view, you know you have twemoji, you don't know from twemoji itself where it came from. The API specifies all configuration options, we had to start from somewhere (as base url)

**Comment by vijayaraghavanramanan** [comment]
@PeterDaveHello 

I don't think there's any way to do this better. 

Twemoji has to provide the basepath as default, so that it is convenient for the user. 

At the same time, it also allows loading of png/svg from other places. Can't think of any way of doing something better than this.

**Comment by WebReflection** [comment]
There is a possible improvement the library could implement.
The base path could be selected through an operation similar to:

```js
base: (
  document.getElementById('twemoji-js') ||
  {src: ''}
).src.replace(/twemoji(?:\.min)?\.js/, '')
```
using the script itself to understand where the script is coming from.
This would setup before other scripts can use `twemoji` the right base, still supporting configuration later on when/if needed.

However, this is inconsistent because it works only if the library is injected as HTML `<script>`, excluding such configuration when import through packagers (browserify, others), excluding a server-side only usage as text replacer, which is also one of the library use cases.

A change like this is surely a minor, hopefully done in a way that never breaks what's using latest 2, and it needs quite some testing.

I'll think about it, maybe it's worth it as fix for the HTTP/2 case which I agree should be considered, it just doesn't fit in a "_work out of the box_" library with remote assets like this one.

**Comment by PeterDaveHello** [comment]
Sorry that I didn't trace the code yet, so I'm not sure if it's a problem, but I'm wondering that why not using relative path instead of absolute path so that there won't be any problem?

**Comment by WebReflection** [comment]
if I use relative path for assets in a script included remotely it will look for these assets from the folder the page included it ... right? Assets must be downloaded from the CDN which is an external source, not a local (relative) one.

**Comment by PeterDaveHello** [comment]
@WebReflection hmmmm ...  not sure if I saw some libs have that ability to use relative path on CDN before, you may be right.

**Comment by WebReflection** [comment]
The script itself doesn't know where it comes from and relative paths in a page are relative to the current page URL.

If there is a way to tell the script to use its domain as root folder then please let me know so I can feature detect that and eventually "save from x-domain fragmentation" at least evergreen/modern browsers that implement such feature.

**Comment by MattIPv4** [comment]
Issue appears to be resolved, closing.
